### ISSUE TITLE
Datastore: Indexed as false doesn't get transmitted (i.e. remains true gcd side)

### ISSUE THREAD

**Issue by gmlewis** [google/google-api-go-client#54]
**peterke** on 14 Jan 2014 at 10:18:

```
Setting the Indexed field of a property to false will result in it being 
ommited from the generated JSON request (due to it being defined as 
"omitempty"), and as such, the default "indexed" value will be used server 
side, which will be true most of the time (e.g. for strings).

Removing the omitempty solves the issue:
Indexed bool `json:"indexed"`

What version of the product are you using? On what operating system?
datastore v1beta2
```

**Comment by gmlewis** [comment]
Successfully auto-migrated from code.google.com:
https://code.google.com/p/google-api-go-client/issues/detail?id=54

**Comment by gmlewis** [comment]
I'm starting to wonder if we should start blacklisting problematic fields and remove the `omitempty` tag for just those.  Leaving this issue open.

**Comment by ppg** [comment]
Isn't this really an issue with all fields, but especially bool fields?  There is no way to send a `false` value currently, it will always omit it, so the API will never get it.  This means you could never unsuspend a user for instance, or any action that requires passing along an explicit `false` in the JSON.

**Comment by gmlewis** [comment]
Please check out https://github.com/google/google-api-go-client/commit/d4ab0b854f8606d689612515c3d18bd3b19d7e70 and see if that solves the problem for you.

**Comment by ppg** [comment]
@gmlewis it doesn't appear to, because the service description for say the `'suspended'` field on admin/directory_v1/user doesn't have a default; ergo its not caught by that change, and doesn't become a pointer, and false is always omitted on JSON generation; similarly if the default in the service description was `'false'`, you'd also have no way to set back to false after you had set something to true.

Basically any bool field that isn't declared as `*bool` but still has omitempty (which they all pretty much do) can never, ever be set to `false` through the Go generated code.  And actually, even more generally, any structs field that isn't a pointer can never have that fields zero value sent through the Go generated code.  Whether there's a default or not is immaterial to the issue I believe.

**Comment by erickoledadevrel** [comment]
Adding on to what @ppg said, d4ab0b8 is not a complete solution, since it only fixes the problems for fields with defaults defined in the schema. There are times when you'll want to empty out a field that didn't have a default, but currently has some value assigned. Currently this doesn't seem possible using the generated code.

Are there any workarounds to this problem that don't involve changes to the generated code? For instance, using a custom struct that has different JSON marshalling rules?

**Comment by ppg** [comment]
@erickoledadevrel Unfortunately I think we're limited by the fields being Go value types; i.e. a `string` is going to always have a value of some sort, so whatever one signified to mean 'not present' will always prevent actually using that value, whether its `''`, `'<not-present>'` or whatever.  That seems to be reflected in how say gRPC has generated the server and client framework code to always use pointers and expose some helper methods to get at things you know are there (i.e. required in the proto specification).

I guess as an alternative you could have a 'present' bool field for all values and have that control the serialization; i.e.
```
user := new(admin.User)
json.Unmarshal([]byte("{}"), user)
fmt.Println(user.SuspendedPresent) # Is false, meaning user.Suspended shouldn't be used
fmt.Println(user.Suspended)        # Still prints false

user.SuspendedPresent = true
user.Suspended = false
out, _ := json.Marshal(user)
fmt.Println(out)                   # {"suspended": false}
```

That seems pretty heavy handed to me though :/

@gmlewis I have a PR linked in this conversation thread that was closed and requested to be put into Gerrit; however it only has the fix to make bool pointers, since I had a specific need for that right now.  Did you guys want to potentially review just that change (in which case I'd send through Gerrit), or did you want to address the issue more holistically on all fields?

**Comment by gmlewis** [comment]
Thanks, @ppg. It will be up to @bradfitz to decide if he wants all bools to be handled as pointers.
I'll let him answer you directly.

**Comment by erickoledadevrel** [comment]
@ppg, Thanks for the background, certainly a sticky problem. I'm an extremely novice Go user, and I don't know which would be more in line with Go conventions, so I'll leave that to the pros. But some solution to this problem is needed, and it may be worth considering fixing the problem for types other than bools as well.

**Comment by fredr** [comment]
This is also a problem with ints (and I guess floats and strings have the same problem), i.e in the Google Cloud Monitoring API its not possible to send custom metrics with value 0, because `cloudmonitoring.Point.Int64Value` has the omitempty flag.

**Comment by ppg** [comment]
Any updates or thoughts on this? Its almost certainly going to require a breaking API change to fully, correctly support all values to the API through this client, so it'd be nice to arrive at a path sooner rather than later.

**Comment by mcgreevy** [comment]
There are a number of subtle issues at play here that make it difficult to create a fix that works in all cases.  For starters, only a subset of fields are affected.  Which ones depends on the server-side implementation of the API.

In the normal case, even if a field is omitted from the JSON that is sent to the server, it will be correctly interpreted.  For example, in the calendar API, the Event resource has a GuestsCanModify boolean field.  If this is left unset as part of an update, the field will be cleared on the server, as update semantics dictate.  If it is left unset as part of a patch, the field is left unchanged, as patch semantics dictate.

Unfortunately, some APIs explicitly check for the presence of some fields, and these are the ones that are being discussed in this issue. Because this issue only affects some fields, we don't want to make a breaking change which affects other fields for no benefit.

Another confounding factor is that the same JSON marshaling code is used for both patch and update operations, which makes tweaking the JSON output more difficult (forcing particular fields to be always sent regardless of their contents works fine for updates, but breaks patches).

I've just sent out a change for @gmlewis to review (https://code-review.googlesource.com/#/c/3382/) which contains point fixes for most of the problematic fields mentioned in this issue and issues linked from it.  The fix is to make just those fields pointers so that it is possible to distinguish between unset and empty.  

I'll be following up soon with another change to address related issues with patch requests.

**Comment by mcgreevy** [comment]
I've submitted fixes for the fields mentioned in this issue (and referenced issues).  These fields are now pointers and so can be explicitly set to the address of any value, including 0, false, etc., and these values will be included in requests sent to the server.

A set of functions have been added to the googleapi package to make it easier to set these pointer fields:

func Bool(v bool) *bool
func Int32(v int32) *int32
func Int64(v int64) *int64
func Float64(v float64) *float64
func Uint32(v uint32) *uint32
func Uint64(v uint64) *uint64
func String(v string) *string

See https://godoc.org/google.golang.org/api/googleapi#Bool

**Comment by catsby** [comment]
Hello! 

I see this issue is closed and the latest discussion was over a year and half ago, but I'd like to know if anyone has thoughts on migrating all(?) fields to use pointers. I'm pretty new around this repo and I just [commented yesterday on working with `bool`s](https://github.com/google/google-api-go-client/issues/201#issuecomment-302151199), specifically I can't send `false` without first listing the field in `ForceSendFields`. 

I realize changing this repo to use pointers is not only a nontrivial task for maintainers, and likely a very backwards incompatible change, but as it stands there's more burden placed on developers using the SDK to manually specify all the fields needed in `ForceSendFields`. (I also didn't see `ForceSendFields` documented, other than mentioned in generated [the godocs](https://godoc.org/google.golang.org/api/storage/v1)). 

So with this issue being closed for some time, and being routed here from other issues for "discussing here", I'm curious if any maintainers have given this switch any thought lately. 

Thanks!

**Comment by jba** [comment]
I don't think we'll migrate all fields to pointers&mdash;that would break too much code&mdash;but we continue to maintain a list of fields that should be pointers. If you notice such fields, let us know by filing an issue.

**Comment by catsby** [comment]
@jba That's understandable I suppose. I was hoping that, because some of these API files are names things like `v1beta3`, it would give ample opportunity to make this change and allow downstream users to adjust. It's beta, after all? 

Either way, I understand if this isn't something the maintainers are willing to do en masse. Would you consider doing this for all `bool` fields though? As `StorageAutoResize` demonstrates, users need to be able to send `false`. Of course, I'm not experienced with this SDK, so I would be interested to learn the parts where you would not want to or expect to be able to send `false`. 

Thanks for the consideration!

**Comment by jba** [comment]
Not all bool fields, but all where false != omitted. Hopefully that's a small subset.

Rather than do a wholesale review, we'll address each case as it arises.
